Sparse matrix: fix fast implementation of findnext and findprev for cartesian coordinates

tkluck opened the pull request:
#23317 introduced special cases for `findnext` in cases of sparse vectors/matrices. The case of matrices was inadvertently broken when `CartesianIndex` was introduced. This pull request fixes that.

In addition, a recent pull request (#31354) identified a need for `findprev(f, m)` to have a sparse implementation not only for the case `f == !iszero`, but also for any `f` that has `f(zero(eltype(m)))` equal to `false`. In particular, the `hash` function is a heavy user of `findprev(!isequal(x), m)` for different values of `x`, including `x == 0`. This is part of this branch, as well.

I feel incredibly bad about submitting this pull request here as one of its commits is a revert of #31354 . These are my reasons for taking that step:

 - In the case of `eltype(m) == BigInt` (or any other `eltype` that exists on the heap), the test for `f(zero(...))` does an allocation which quickly becomes a major bottleneck.
 - fixing/extending the existing logic from #23317 seems more sensible than duplicating functionality.

I did make a point of including the new test cases from the other pull request. For example, that ensures that these functions work with
```julia
w = [ "a" ""; "" "b"]
w_sp = sparse(w)
```
as their argument.

tkluck committed:
Revert "sparse findnext findprev hash performance improved (#31354)"

This seems to duplicate work from #23317 and it causes performance
degradation in the cases that one was designed for. See
https://github.com/JuliaLang/julia/pull/31354#issuecomment-491511974

This reverts commit e0bef65d5ac228a0e84e0cee4013c245ce2009eb.

tkluck committed:
_sparse_findnextnz: work with CartesianIndex

KlausC committed:
sparse findXXX: cherry-pick test cases from #31354

tkluck committed:
sparse findprev/next: skip zeros if the predicate is false on zero(eltype(...))

ViralBShah commented:
Pinging @mbauman @andreasnoack for review.

KristofferC commented:
We need to decide what to do for 1.2 here.

Either leave things as is, revert https://github.com/JuliaLang/julia/pull/31354, or merge this and backport to the 1.2 branch.

ViralBShah commented:
Does this maintain the performance improvements for the common case described in #31354? If so, let's merge this, and also add test cases to prevent regression.

tkluck commented:
Yes, it maintains the same behaviour of `hash` time scaling with fill rate.

mbauman commented:
Ah, ok, this is indeed much simpler.  Just to wrap my own head around what happened here:

* We initially just had one `findnext(f::typeof(!iszero), v::AbstractSparseArray, i::Integer)`, which used `_sparse_findnextnz` to do its work.  This was sub-optimal for two reasons: it didn't work for CartesianIndices (after the find-to-keys-ificiation), and it only handled `!iszero`.
* #31354 fixed this by adding an _additional_ `findnext(pred::Function, A::SparseArrays.SparseMatrixCSC, ij::CartesianIndex{2})` method and essentially a parallel `_sparse_findnextnz` system that worked on CartesianIndices instead of Integers.
* This PR fixes up the original system, making `_sparse_findnextnz` work with CartesianIndices.  I thought it should be quicker, too, but it looks like that's not the case in this simple spot-check:

```julia
julia> using Revise, SparseArrays, BenchmarkTools

julia> Revise.track(SparseArrays)

julia> A = spzeros(10000,10000);

julia> A[end,end]= 1
1

julia> @benchmark findnext(isequal(0), $A, CartesianIndex(1,1))
BenchmarkTools.Trial:
  memory estimate:  240 bytes
  allocs estimate:  5
  --------------
  minimum time:     700.908 ns (0.00% GC)
  median time:      752.756 ns (0.00% GC)
  mean time:        812.077 ns (9.16% GC)
  maximum time:     448.030 μs (99.81% GC)
  --------------
  samples:          10000
  evals/sample:     119

julia> @benchmark findnext(!isequal(0), $A, CartesianIndex(1,1))
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     27.390 ns (0.00% GC)
  median time:      29.360 ns (0.00% GC)
  mean time:        28.993 ns (0.00% GC)
  maximum time:     47.038 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     995

shell> git merge --no-commit pr/32007
Automatic merge went well; stopped before committing as requested

julia> @benchmark findnext(isequal(0), $A, CartesianIndex(1,1))
BenchmarkTools.Trial:
  memory estimate:  240 bytes
  allocs estimate:  5
  --------------
  minimum time:     683.640 ns (0.00% GC)
  median time:      736.847 ns (0.00% GC)
  mean time:        791.763 ns (8.20% GC)
  maximum time:     376.179 μs (99.77% GC)
  --------------
  samples:          10000
  evals/sample:     150

julia> @benchmark findnext(!isequal(0), $A, CartesianIndex(1,1))
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     6.755 μs (0.00% GC)
  median time:      6.774 μs (0.00% GC)
  mean time:        6.955 μs (0.00% GC)
  maximum time:     13.100 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     5
```

JeffBezanson commented:
This could use extra test cases to cover what broke.

KristofferC commented:
I think it was only performance?

tkluck commented:
@mbauman yes, your summary sums it up. As for your spot benchmarks -- I identified the reason for the slower performance and am about to add a fix for them to this branch.

tkluck commented on stdlib/SparseArrays/test/sparse.jl:
```diff
@@ -2263,16 +2263,20 @@ end
     end
 
-    y = [0 0 0 0 0;
+    y = [7 0 0 0 0;
```
Agreed! I'm about to push a rebased version that attributes that code to him. I'm also in favour or squashing into a co-authored version.

tkluck committed:
_sparse_findnextnz: search m.colptr using searchsorted for performance

Thanks to @mbauman for spotting this issue in
https://github.com/JuliaLang/julia/pull/32007#issuecomment-492018400.

tkluck commented:
@JeffBezanson and @KristofferC , yes it was only performance.

tkluck commented on stdlib/SparseArrays/src/sparsematrix.jl:
```diff
@@ -1312,36 +1312,34 @@ function findnz(S::SparseMatrixCSC{Tv,Ti}) where {Tv,Ti}
     return (I, J, V)
 end
 
-function _sparse_findnextnz(m::SparseMatrixCSC, i::Integer)
```
Indeed, they can be removed and I did.

KristofferC commented:
This wasn't backported to RC2 because it still had the triage label on. I assume that we just go with this PR though.
